sklearn/cluster/_hdbscan/hdbscan.py

```diff
@@ -0,0 +1,895 @@
+
+    min_spanning_tree = mst_linkage_core(mutual_reachability_)
+
+    # Warn if the MST couldn't be constructed around the missing distances
+    if np.isinf(min_spanning_tree.T[2]).any():
+        warn(
+            "The minimum spanning tree contains edge weights with value "
+            "infinity. Potentially, you are missing too many distances "
+            "in the initial distance matrix for the given neighborhood "
+            "size.",
+            UserWarning,
+        )
+
+    return _process_mst(min_spanning_tree)
+
+
+def _hdbscan_sparse_distance_matrix(
+    X,
+    min_samples=5,
+    alpha=1.0,
+    **metric_params,
+):
+    assert issparse(X)
+
+    # Compute sparse mutual reachability graph
+    # if max_dist > 0, max distance to use when the reachability is infinite
+    max_dist = metric_params.get("max_dist", 0.0)
+    mutual_reachability_ = sparse_mutual_reachability(
+        X.tolil(), min_points=min_samples, max_dist=max_dist, alpha=alpha
+    )
+    # Check connected component on mutual reachability
+    # If more than one component, it means that even if the distance matrix X
+    # has one component, there exists with less than `min_samples` neighbors
+    if (
+        csgraph.connected_components(
+            mutual_reachability_, directed=False, return_labels=False
+        )
+        > 1
+    ):
+        raise ValueError(
+            "There exists points with less than %s neighbors. "
+            "Ensure your distance matrix has non zeros values for "
+            "at least `min_sample`=%s neighbors for each points (i.e. K-nn graph), "
+            "or specify a `max_dist` to use when distances are missing."
+            % (min_samples, min_samples)
+        )
+
+    # Compute the minimum spanning tree for the sparse graph
+    sparse_min_spanning_tree = csgraph.minimum_spanning_tree(mutual_reachability_)
+
+    # Convert the graph to scipy cluster array format
+    nonzeros = sparse_min_spanning_tree.nonzero()
+    nonzero_vals = sparse_min_spanning_tree[nonzeros]
+    min_spanning_tree = np.vstack(nonzeros + (nonzero_vals,)).T
+
+    # Sort edges of the min_spanning_tree by weight
+    min_spanning_tree = min_spanning_tree[np.argsort(min_spanning_tree.T[2]), :][0]
+
+    # Convert edge list into standard hierarchical clustering format
+    single_linkage_tree = label(min_spanning_tree)
+
+    return single_linkage_tree
+
+
+def _hdbscan_prims(
+    X,
+    algo,
+    min_samples=5,
+    alpha=1.0,
+    metric="euclidean",
+    leaf_size=40,
+    n_jobs=4,
+    **metric_params,
+):
+    # The Cython routines used require contiguous arrays
+    if not X.flags["C_CONTIGUOUS"]:
+        X = np.array(X, order="C")
+
+    # Get distance to kth nearest neighbour
+    nbrs = NearestNeighbors(
+        n_neighbors=min_samples,
+        algorithm=algo,
+        leaf_size=leaf_size,
+        metric=metric,
+        metric_params=metric_params,
+        n_jobs=n_jobs,
+        p=None,
```

Comment:
Why is `p=None` here?

Reply:
To avoid the `SyntaxWarning` when passing `p` through `metric_params`. The alternative would be to pop `p` out of `metric_params` before this and pass it directly in `__init__` but I preferred this stylistically. Open to changing it if wanted.

Reply:
`NearestNeighbors` directly checks for `p is None` and is validated as such, so I think we should be fine here